### PR TITLE
[Android] Add script to build Android API

### PR DESCRIPTION
**Pull request by jaeyun-jung**
Add script to build Android API library.

To build NNStreamer Android API, pre-built libraries (gstreamer and tf-lite) and gradle build-scripts are needed.
When building the API library, pre-downloaded files are copied into build directory and run gradle-build.

**Changelogs**
* Version 4:
  - update prerequisites and dependency with review comment

* Version 3:
  - limit api module to build android library

* Version 2:
  - remove sudo in the script

* Version 1:
  - initial commit to add script for android api

Signed-off-by: Jaeyun Jung <jy1210.jung@samsung.com>

### CODE REVIEW

**Comment by jaeyun-jung** [comment]
Waiting for 2 PRs in nnstreamer
https://github.com/nnsuite/nnstreamer/pull/1575
https://github.com/nnsuite/nnstreamer/pull/1558

**Comment by jaeyun-jung** [comment]
Result (Android library):
http://nnsuite.mooo.com/nnstreamer/ci/repo-workers/pr-checker/1558-201908062126190.35261201858521-e3046c70ed03a6e0169f6a56660aad1ed367dfdc/binary_repository/ANDROID/


Build logs:
http://nnsuite.mooo.com/nnstreamer/ci/repo-workers/pr-checker/1558-201908062126190.35261201858521-e3046c70ed03a6e0169f6a56660aad1ed367dfdc/report/build_log_1558_android_api.txt

File: ci/taos/plugins-base/pr-audit-build-android.sh [hunk 1]
```diff
@@ -168,6 +185,42 @@ function pr-audit-build-android-run-queue(){
...
         echo -e "[DEBUG] The current directory: $(pwd)."
 
+        # Start to build NNStreamer API for Android.
+        echo -e "[DEBUG] Starting gradle build for Android API."
+
+        # NNStreamer root directory for build.
+        export NNSTREAMER_ROOT=$(pwd)
+
+        # Make directory to build NNStreamer library.
+        sudo mkdir -p build_android_lib
```
Comment (myungjoo) [inline ci/taos/plugins-base/pr-audit-build-android.sh]: Is sudo really OK?
Is sudo really required?

Comment (jaeyun-jung) [reply ci/taos/plugins-base/pr-audit-build-android.sh]: Changed permission of this script on aws (root > www) and removed sudo.
Now running build test in AWS.

File: ci/taos/plugins-base/pr-audit-build-android.sh [hunk 2]
```diff
@@ -57,7 +57,23 @@
 # vi ~/.bashrc
 # # gst-android prebuilt binary (e.g., .a, .so, .h)
 # export GSTREAMER_ROOT_ANDROID=/var/www/html/android/gst_root_android/
+#----------------------------------------------------------------------
```
Comment (leemgs) [inline ci/taos/plugins-base/pr-audit-build-android.sh]: @jaeyun-jung , Please keep the writing style of the existing manual by checking the "Prerequisites" statement.
For example, 

* Before : Step 1/2 , 2/2
* After   : Step 1/4 , 2/4 , 3/4, 4/4

File: ci/taos/plugins-base/pr-audit-build-android.sh [hunk 3]
```diff
@@ -98,6 +114,7 @@ function pr-audit-build-android-run-queue(){
     check_dependency sudo
     check_dependency curl
     check_dependency ndk-build
+    check_dependency gradle
```
Comment (leemgs) [inline ci/taos/plugins-base/pr-audit-build-android.sh]: @jaeyun-jung  Please append "check_dependency sed" statement.

File: ci/taos/plugins-base/pr-audit-build-android.sh [hunk 4]
```diff
@@ -168,6 +185,42 @@ function pr-audit-build-android-run-queue(){
...
+        sed -i "s|nnstreamerRoot=nnstreamer-path|nnstreamerRoot=$NNSTREAMER_ROOT|" build_android_lib/gradle.properties
+
+        # Build Android API.
+        pushd ./build_android_lib
+        sh ./gradlew api:assembleRelease 1>> ../../report/build_log_${input_pr}_android_api.txt
+        popd
+
+        # Check if build procedure is done.
+        nnstreamer_android_api_lib=./build_android_lib/api/build/outputs/aar/api-release.aar
+        if [[ -e $nnstreamer_android_api_lib ]]; then
```
Comment (leemgs) [inline ci/taos/plugins-base/pr-audit-build-android.sh]: Looks good to me at least.

File: ci/taos/plugins-base/pr-audit-build-android.sh [hunk 5]
```diff
@@ -168,6 +185,42 @@ function pr-audit-build-android-run-queue(){
...
+        export NNSTREAMER_ROOT=$(pwd)
+
+        # Make directory to build NNStreamer library.
+        mkdir -p build_android_lib
+
+        # Copy the files (native and java to build Android library) to build directory.
+        cp -r ./api/android/* ./build_android_lib
+
+        # Get the prebuilt libraries and build-script.
+        cp -r /var/www/html/android/nnstreamer-android-api/* ./build_android_lib
```
Comment (leemgs) [inline ci/taos/plugins-base/pr-audit-build-android.sh]: Could you append the variable $ROOT_ANDROID_CI  to remove duplicated folder names such as "/var/www/html/android"?

For example, 
```bash
ROOT_ANDROID_CI="/var/www/html/android"
cp -r ${ROOT_ANDROID_CI}/nnstreamer-android-api/* ./build_android_lib
```

File: ci/taos/plugins-base/pr-audit-build-android.sh (hunk 3 above)
Comment (jaeyun-jung) [reply ci/taos/plugins-base/pr-audit-build-android.sh]: now added a dependency 'sed'

File: ci/taos/plugins-base/pr-audit-build-android.sh (hunk 2 above)
Comment (jaeyun-jung) [reply ci/taos/plugins-base/pr-audit-build-android.sh]: changed to 'Step #/4:'

File: ci/taos/plugins-base/pr-audit-build-android.sh (hunk 5 above)
Comment (jaeyun-jung) [reply ci/taos/plugins-base/pr-audit-build-android.sh]: As you commented, I added ROOT_ANDROID_CI and changed common path.

File: ci/taos/plugins-base/pr-audit-build-android.sh [hunk 6]
```diff
@@ -82,22 +96,27 @@ function pr-audit-build-android-run-queue(){
...
+    export GSTREAMER_ROOT_ANDROID=$ROOT_ANDROID_CI/gst_root_android/
     echo "Exporting a GSTREMER_ROOT_ANDROID path..."
     echo $GSTREAMER_ROOT_ANDROID
 
     # Check if dependent packages are installed. Please add required packages here.
     check_dependency sudo
     check_dependency curl
     check_dependency ndk-build
+    check_dependency sed
+    check_dependency gradle
```
Comment (leemgs) [inline ci/taos/plugins-base/pr-audit-build-android.sh]: Is "**gradle**" command correct? 
As we know, we use "**graldew**" command to build ***.apk** on the terminal.

Comment (jaeyun-jung) [reply ci/taos/plugins-base/pr-audit-build-android.sh]: gradlew is a shell script, not a build command.

File: ci/taos/plugins-base/pr-audit-build-android.sh [hunk 7]
```diff
@@ -168,6 +187,42 @@ function pr-audit-build-android-run-queue(){
         popd
         echo -e "[DEBUG] The current directory: $(pwd)."
 
+        # Start to build NNStreamer API for Android.
```
Comment (leemgs) [inline ci/taos/plugins-base/pr-audit-build-android.sh]: @jaeyun-jung 
When you think the PR is a final version, Please append the execution result of the below statement in PR body. 
```bash
u1604$ sudo apt install shellcheck
u1604$ shellcheck  ./pr-audit-build-android.sh
```

File: ci/taos/plugins-base/pr-audit-build-android.sh (hunk 6 above)
Comment (leemgs) [reply ci/taos/plugins-base/pr-audit-build-android.sh]: Sure, **gradlew** is a shell script. and **gradle** is a command that is written by JAVA. So, I wonder about the "check_dependency gradle" statement if this statement is appropriate or not

Comment (jaeyun-jung) [reply ci/taos/plugins-base/pr-audit-build-android.sh]: I understand. now we are using the studio and gradle is unnecessary.
Now I updated the PR to remove dependency.

**Comment by jaeyun-jung** [comment]
shellcheck result:


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 78:
    message="Trigger: wait queue. There are other build jobs and we need to wait.. The commit number is $input_commit."
                                                                                                        ^-- SC2154: input_commit is referenced but not assigned.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 79:
    cibot_report $TOKEN "pending" "TAOS/pr-audit-build-android" "$message" "${CISERVER}${PRJ_REPO_UPSTREAM}/ci/${dir_commit}/" "$GITHUB_WEBHOOK_API/statuses/$input_commit"
                 ^-- SC2086: Double quote to prevent globbing and word splitting.
                                                                                                               ^-- SC2154: dir_commit is referenced but not assigned.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 85:
    cibot_report $TOKEN "pending" "TAOS/pr-audit-build-android" "$message" "${CISERVER}${PRJ_REPO_UPSTREAM}/ci/${dir_commit}/" "$GITHUB_WEBHOOK_API/statuses/$input_commit"
                 ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 91:
    cibot_report $TOKEN "pending" "TAOS/pr-audit-build-android" "$message" "${CISERVER}${PRJ_REPO_UPSTREAM}/ci/${dir_commit}/" "$GITHUB_WEBHOOK_API/statuses/$input_commit"
                 ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 106:
    echo $PATH
         ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 136:
        $result=999
         ^-- SC1066: Don't use $ on the left side of assignments.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 156:
        time ndk-build NDK_PROJECT_PATH=. APP_BUILD_SCRIPT=./Android-nnstreamer.mk NDK_APPLICATION_MK=./Application.mk -j$(nproc) \
                                                                                                                         ^-- SC2046: Quote this to prevent word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 157:
        2> ../../report/build_log_${input_pr}_android_error.txt \
                                  ^-- SC2154: input_pr is referenced but not assigned.
                                  ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 158:
        1> ../../report/build_log_${input_pr}_android_output.txt
                                  ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 159:
        result=$(($result+$?))
                  ^-- SC2004: $/${} is unnecessary on arithmetic variables.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 164:
        time ndk-build NDK_PROJECT_PATH=. APP_BUILD_SCRIPT=./Android-app.mk NDK_APPLICATION_MK=./Application.mk -j$(nproc) \
                                                                                                                  ^-- SC2046: Quote this to prevent word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 165:
        2>> ../../report/build_log_${input_pr}_android_error.txt \
                                   ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 166:
        1>> ../../report/build_log_${input_pr}_android_output.txt
                                   ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 167:
        result=$(($result+$?))
                  ^-- SC2004: $/${} is unnecessary on arithmetic variables.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 176:
            mkdir -p ../$PACK_BIN_FOLDER/ANDROID
                        ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 177:
            sudo mv ../out ../$PACK_BIN_FOLDER/ANDROID/
                              ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 193:
        export NNSTREAMER_ROOT=$(pwd)
               ^-- SC2155: Declare and assign separately to avoid masking return values.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 209:
        sh ./gradlew api:assembleRelease 1>> ../../report/build_log_${input_pr}_android_api.txt
                                                                    ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 216:
            mkdir -p ../$PACK_BIN_FOLDER/ANDROID
                        ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 217:
            cp $nnstreamer_android_api_lib ../$PACK_BIN_FOLDER/ANDROID/nnstreamer-api.aar
                                              ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 220:
            result=$(($result+2))
                      ^-- SC2004: $/${} is unnecessary on arithmetic variables.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 230:
    time_diff=$(($time_end-$time_start))
                 ^-- SC2004: $/${} is unnecessary on arithmetic variables.
                           ^-- SC2004: $/${} is unnecessary on arithmetic variables.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 231:
    time_build_cost="$(($time_diff / 60))m $(($time_diff % 60))s"
                        ^-- SC2004: $/${} is unnecessary on arithmetic variables.
                                              ^-- SC2004: $/${} is unnecessary on arithmetic variables.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 242:
        cibot_report $TOKEN "success" "TAOS/pr-audit-build-android" "$message" "${CISERVER}${PRJ_REPO_UPSTREAM}/ci/${dir_commit}/" "$GITHUB_WEBHOOK_API/statuses/$input_commit"
                     ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 245:
        cibot_report $TOKEN "success" "(INFO)TAOS/pr-audit-all" "$message" "${CISERVER}${PRJ_REPO_UPSTREAM}/ci/${dir_commit}/" "$GITHUB_WEBHOOK_API/statuses/$input_commit"
                     ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 258:
                global_check_result="failure"
                ^-- SC2034: global_check_result appears unused. Verify it or export it.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 264:
            cibot_report $TOKEN "success" "TAOS/pr-audit-build-android" "$message" "${CISERVER}${PRJ_REPO_UPSTREAM}/ci/${dir_commit}/" "$GITHUB_WEBHOOK_API/statuses/$input_commit"
                         ^-- SC2086: Double quote to prevent globbing and word splitting.


In ./ci/taos/plugins-base/pr-audit-build-android.sh line 267:
            cibot_report $TOKEN "failure" "TAOS/pr-audit-build-android" "$message" "${CISERVER}${PRJ_REPO_UPSTREAM}/ci/${dir_commit}/" "$GITHUB_WEBHOOK_API/statuses/$input_commit"
                         ^-- SC2086: Double quote to prevent globbing and word splitting.

File: ci/taos/plugins-base/pr-audit-build-android.sh (hunk 7 above)
Comment (jaeyun-jung) [reply ci/taos/plugins-base/pr-audit-build-android.sh]: added as a comment

**Comment by myungjoo** [comment]
shellcheck result gives you a lot of potential errors... :(

**Pull request merged**